### PR TITLE
Compute hdiv norm 2

### PR DESCRIPTION
**Pull request by bangerth**
This does what #831 was supposed to do.

### CODE REVIEW

**Comment by tjhei** [comment]
You compute it using the components ```0..dim```, which means it is not very flexible (assumes the vector part starts at component 0). Another option would be to sum over all components and let the user pick the right components using the weight function. Thoughts?

**Comment by bangerth** [comment]
I thought how to do that. The function in which it is called does not take an argument that allows us to select components, and if we get a weight there is no way of querying what the *intention* of the weight is either (i.e., I can't find out whether the weight selects exactly dim components with weight 1).

So, I think there is no good way of making it more generic. On the other hand, all of the examples we typically have happen to put the velocity or displacement fields first, so for the majority of cases this seems to be enough.

**Comment by tjhei** [comment]
Using ```for (unsigned int k=0; k<n_components; ++k)``` would do the right thing, or not?

I don't understand your argument. You don't want to sum all components because you don't know how to check if the user indeed selected exactly dim components? You could argue the same for the H1 norm...

**Comment by bangerth** [comment]
The gradient of a vector function is a matrix of size `n_components` by `dim`. I can take the Frobenius norm of this matrix, and that is exactly what the function does when you compute the H1 seminorm.

But the divergence of a vector function is only defined if `n_components` is *equal to* `dim`. Otherwise, I can't take the trace of such a matrix. As an extension, I allow the case where `n_components > dim`, in which case I take the trace of the `dim x dim` matrix formed by the  first `dim` rows.

**Comment by tjhei** [comment]
Doh, silly me. Of course.

File: tests/hp/integrate_difference_03.cc [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  // u(x)=x, so the norm must be equal to 1
+  // on every cell
+  {
+    VectorTools::integrate_difference (dof_handler,
+                                       vec,
+                                       ZeroFunction<dim>(dim),
+                                       diff,
+                                       q_collection,
+                                       VectorTools::W1infty_seminorm);
+    deallog << "Hdiv semi, diff=" << diff.linfty_norm() << std::endl;
```
Comment (tjhei) [inline tests/hp/integrate_difference_03.cc]: why is this output not in the ```.output``` file?

**Comment by bangerth** [comment]
Good catch -- thanks for spotting it. I must have forgotten to commit the changed output as well. I guess I really screwed up with this whole patch...

File: tests/hp/integrate_difference_03.cc [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  // interpolate a linear function
+  Vector<double> vec (dof_handler.n_dofs());
+  VectorTools::interpolate (dof_handler,
+			    VectorFunctionFromScalarFunctionObject<dim> (&f<dim>, 0, dim),
+                            vec);
+
+  Vector<float> diff (tria.n_active_cells());
+
+  // H1 seminorm. the function is u(x)=x, so
+  // its H1 seminorm should be equal to 1/2
```
Comment (tjhei) [inline tests/hp/integrate_difference_03.cc]: 0.5? the output is 1...

File: tests/hp/integrate_difference_03.cc [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                                       vec,
+                                       ZeroFunction<dim>(dim),
+                                       diff,
+                                       q_collection,
+                                       VectorTools::W1infty_seminorm);
+    deallog << "Hdiv semi, diff=" << diff.linfty_norm() << std::endl;
+    // also ensure that we indeed get the
+    // same value on every cell
+    diff. add(-1);
+    AssertThrow (diff.l2_norm() == 0, ExcInternalError());
```
Comment (tjhei) [inline tests/hp/integrate_difference_03.cc]: whoa, that works? I was expecting you need ```< eps```.

File: tests/hp/integrate_difference_03.cc (hunk 2 above)
Comment (bangerth) [inline tests/hp/integrate_difference_03.cc]: Yes, copy-paste error. Will fix this here and in the place where it came from in a separate, follow-up patch.

File: tests/hp/integrate_difference_03.cc (hunk 3 above)
Comment (bangerth) [inline tests/hp/integrate_difference_03.cc]: Yea, fascinating, isn't it?

**Comment by tjhei** [comment]
Okay, please merge yourself (optionally fixing the comment before you do).

**Comment by bangerth** [comment]
Done. I fixed the comments.

**Pull request merged**